src/cgw_connection_processor.rs

```diff
@@ -264,48 +304,62 @@ impl CGWConnectionProcessor {
 
-            // TODO: refactor
-            // Round-robin selection of stream to process:
-            // first, get single message from WSS, then get a single msg from RX MBOX
-            // It's done to ensure we process WSS and RX MBOX equally with same priority
-            // Also, we have to make sure we don't sleep-wait for any of the streams to
-            // make sure we don't cancel futures that are used for stream processing,
-            // especially TCP stream, which is not cancel-safe
-            if poll_wss_first {
-                if let Some(val) = stream.next().now_or_never() {
-                    if let Some(res) = val {
-                        if let Ok(msg) = res {
-                            wakeup_reason = WakeupReason::WSSRxMsg(Ok(msg));
-                        } else if let Err(msg) = res {
-                            wakeup_reason = WakeupReason::WSSRxMsg(Result::Err(msg));
+                match queue_lock.get_device_queue_state(&device_mac).await {
+                    CGWUCentralMessagesQueueState::Rx | CGWUCentralMessagesQueueState::RxTx => {
+                        // Check if message Queue is not empty
+                        if let Some(queue_msg) =
+                            queue_lock.dequeue_device_message(&device_mac).await
+                        {
+                            // Get message from queue
+                            pending_req_id = queue_msg.parsed_message.id;
+                            pending_req_type = queue_msg.parsed_message.cmd_type.clone();
+                            wakeup_reason = WakeupReason::MboxRx(Some(
+                                CGWConnectionProcessorReqMsg::SinkRequestToDevice(queue_msg),
+                            ));
+                            debug!("Got pending request with id: {}", pending_req_id);
+                            if pending_req_type == CGWUCentralCommandType::Factory
+                                || pending_req_type == CGWUCentralCommandType::Upgrade
+                            {
+                                queue_lock
+                                    .set_device_queue_state(
+                                        &device_mac,
+                                        CGWUCentralMessagesQueueState::Discard,
+                                    )
+                                    .await;
+                            } else if pending_req_type == CGWUCentralCommandType::Reboot {
+                                queue_lock
+                                    .set_device_queue_state(
+                                        &device_mac,
+                                        CGWUCentralMessagesQueueState::Rx,
+                                    )
+                                    .await;
+                            }
+
+                            fsm_state = CGWUCentralMessageProcessorState::Pending;
+                        } else if let Some(val) = mbox_rx.recv().now_or_never() {
```

Comment:
The mbox_rx has the highest priority.
 it should most likely be put at the very top ~after line 314;
 
 In your current impl, the mbox_rx has almost lowest prio:
 * fetched only if device is in state pending
 * only if queue is empty;
 
 What if a duplicate connection's been seen, and this connection must be closed immediately? While this connection processor would still alterate the queue?